corehq/motech/repeaters/expression/forms.py

```diff
@@ -39,3 +70,39 @@ def clean_configured_filter(self):
             raise ValidationError(e)
 
         return self.cleaned_data['configured_filter']
+
+    def clean_case_action_expression(self):
+        raw = self.cleaned_data.get('case_action_expression')
+        if raw:
+            try:
+                ExpressionFactory.from_spec(
+                    raw, FactoryContext.empty(domain=self.domain)
+                )
+            except BadSpecError as e:
+                raise ValidationError(e)
+
+        return raw
+
+    def clean_case_action_filter_expression(self):
+        raw = self.cleaned_data.get('case_action_filter_expression')
+        if raw:
+            try:
+                FilterFactory.from_spec(raw)
+            except BadSpecError as e:
+                raise ValidationError(e)
+
+        return raw
+
+    def clean(self):
+        cleaned_data = super().clean()
+        case_filter = bool(cleaned_data.get('case_action_filter_expression'))
+        case_operation = bool(cleaned_data.get('case_action_expression'))
+        if case_filter ^ case_operation:
+            field = 'case_action_expression' if case_filter else 'case_action_filter_expression'
+            other = 'case_action_filter_expression' if case_filter else 'case_action_expression'
+            raise ValidationError({
+                field: _("This field is required when '{other}' is provided").format(
```

Comment:
nit: 
I might be missing something here so just checking if this is possible in the respective `clean_` methods for these properties i.e a `ValidationError` is raised if the property itself if not present while the other is?

Reply:
you can't guarantee order in the individual `clean_x` methods so any cross validation needs to be done in the main `clean` method.